app/models/megaphone/publisher.rb

```diff
@@ -0,0 +1,133 @@
+module Megaphone
+  class Publisher < Integrations::Base::Publisher
+    WAIT_INTERVAL = 5.seconds
+    WAIT_TIMEOUT = 5.minutes
+
+    attr_reader :feed
+    attr_accessor :wait_interval
+
+    def initialize(feed)
+      @feed = feed
+      @wait_interval = WAIT_INTERVAL
+    end
+
+    def megaphone_podcast
+      @megaphone_podcast ||= Megaphone::Podcast.find_by_feed(feed)
+    end
+
+    alias_method :show, :megaphone_podcast
+    alias_method :private_feed, :feed
+
+    def publish!
+      sync_podcast!
+      sync_episodes!
+    end
+
+    def sync_episodes!
+      Rails.logger.tagged("Megaphone::Publisher#sync_episodes!") do
+        # delete or unpublish episodes we aren't including in the feed anymore
+        unpublish_and_delete_episodes!
+
+        # start with create and update, make sure they have been created at least
+        create_and_update_episodes!
+
+        # check if the upload has completed and the audio has finished processing
+        check_status_episodes!
+      end
+    end
+
+    # For status
+    # :uploaded = the latest media version was ready on dtr, and was saved to mp
+    # :delivered = attributes saved, media finished processing
+    # if not uploaded, and media ready, try to set that on the next update
+    # if uploaded and not delivered, check mp status, see if processing done
+    def check_status_episodes!
+      episodes = private_feed.episodes.unfinished(:megaphone)
+      timeout_at = Time.now.utc + WAIT_TIMEOUT
+
+      while episodes.size > 0 && Time.now.utc < timeout_at
+        sleep(wait_interval)
+        episodes = check_episodes(episodes)
+      end
+
+      # if after all those checks, still incomplete? throw an error
+      if episodes.size > 0
+        msg = "Megaphone::Publisher.check_status_episodes! timed out on: #{episodes.map(&:id)}"
+        Logger.error(msg)
+        raise msg
+      end
+    end
+
+    def check_episodes(episodes)
+      remaining = []
+
+      episodes.each do |ep|
+        megaphone_episode = Megaphone::Episode.find_by_episode(megaphone_podcast, ep)
+
+        # check if it is uploaded yet
+        # if not go looking for the DTR media version
+        if !megaphone_episode.delivery_status.uploaded?
+          megaphone_episode.upload_audio!
+        end
+
+        # check if it is uploaded, but not delivered - see if megaphone has processed
+        status = megaphone_episode.delivery_status
+        if !status.delivered? && status.uploaded?
+          megaphone_episode.check_audio!
+        end
+
+        if !ep.episode_delivery_status(:megaphone).delivered?
+          remaining << ep
+        end
+      end
+    end
+
+    def create_and_update_episodes!
+      megaphone_episodes = []
+      private_feed.episodes.unfinished(:megaphone).each do |ep|
```

Comment:
Do you see the megaphone integration as operating independently from the logic in this [abstract episode set operations concern](https://github.com/PRX/feeder.prx.org/pull/1112/files#diff-92e7117a8391a2f8207842a2c3ac800f2d62db2e82150c9270b95426f5d5da66)?

With the approach here, we have a simpler/compact interface to the feed's episodes, with no shared episode set concerns between the two existing integrations.  But we're lacking visibility into deleted episodes, sorting properties etc. Does that matter?

Reply:
The episode set operations are used in the base publisher in a few ways - 
`episodes_to_sync` doesn't work for megaphone, as it filters to only published episodes.
```
      def episodes_to_sync
        filter_episodes_to_sync(show.episodes)
      end
```
it is also wrapped with `filter_episodes_to_sync`, which after querying and instantiating all the episodes rejects from the array based on delivery status, whereas the `unfinished` method starts with only episodes in need of delivery based on delivery status, which is more efficient, and includes the draft episodes we need to create/update on megaphone.

The archive episodes logic looks closer to what is needed for megaphone, so that method might be useful when that gets implemented.

Reply:
I'll also mention, the actual filter for deciding which episodes should be included in the sync is not really using the episode set operations - in the `show` it's getting the list of `feed_ready` episodes - so this is not useful for the megaphone integration as a super class method, and the querying to get the episode ids to filter the episodes from the episode set operation does not seem like a better way to do this - I'd rather use scopes and do all the filtering as part of the sql query:

https://github.com/PRX/feeder.prx.org/blob/feat/megaphone/app/models/integrations/base/show.rb#L31